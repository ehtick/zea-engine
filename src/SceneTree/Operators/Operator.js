import { sgFactory } from '../SGFactory'
import { ItemFlags, BaseItem } from '../BaseItem.js'

/**
 * Class representing an operator.
 *
 * @extends BaseItem
 */
class Operator extends BaseItem {
  /**
   * Create an operator.
   * @param {string} name - The name value.
   */
  constructor(name) {
    super(name)

    // Items which can be constructed by a user (not loaded in binary data).
    // Should always have this flag set.
    this.setFlag(ItemFlags.USER_EDITED)

    this.__inputs = []
    this.__outputs = []
    this.__evalOutput = this.__evalOutput.bind(this)
  }

  /**
   * This method sets the state of the operator to dirty which propagates
   * to the outputs of this operator, and which may then propagate to other
   * operators. When the scene is cleaned, which usually is caused by rendering
   * then the chain of operators are cleaned by triggering evaluation.
   * @private
   */
  setDirty() {
    for (const o of this.__outputs) o.setDirty()
  }

  /**
   * This method can be overridden in derived classes
   * to perform general updates (see GLPass or BaseItem).
   *
   * @param {object} event
   * @private
   */
  __parameterValueChanged(event) {
    super.__parameterValueChanged(event)
    this.setDirty()
  }

  /**
   * The addInput method.
   * @param {OperatorInput} output - The output value.
   * @return {array} - The return value.
   */
  addInput(input) {
    input.setOperator(this)
    this.__inputs.push(input)
    return input
  }

  /**
   * The removeInput method.
   * @param {OperatorInput} input - The input value.
   */
  removeInput(input) {
    if (input.getParam()) input.getParam().unbindOperator(this)
    this.__inputs.splice(this.__inputs.indexOf(input), 1)
  }

  /**
   * Getter for the number of inputs in this operator.
   * @return {number} - Returns the number of inputs.
   */
  getNumInputs() {
    return this.__inputs.length
  }

  /**
   * The getInputByIndex method.
   * @param {number} index - The index value.
   * @return {object} - The return value.
   */
  getInputByIndex(index) {
    return this.__inputs[index]
  }

  /**
   * The getInput method.
   * @param {string} name - The name value.
   * @return {OperatorInput} - The return value.
   */
  getInput(name) {
    for (const o of this.__inputs) {
      if (o.getName() == name) return o
    }
  }

  /**
   * The addOutput method.
   * @param {OperatorOutput} output - The output value.
   * @return {array} - The return value.
   */
  addOutput(output) {
    output.setOperator(this)
    this.__outputs.push(output)
    return output
  }

  /**
   * The removeOutput method.
   * @param {OperatorOutput} output - The output value.
   */
  removeOutput(output) {
    if (output.getParam()) output.getParam().unbindOperator(this)
    this.__outputs.splice(this.__outputs.indexOf(output), 1)
  }

  /**
   * Getter for the number of outputs in this operator.
   * @return {number} - Returns the number of outputs.
   */
  getNumOutputs() {
    return this.__outputs.length
  }

  /**
   * The getOutputByIndex method.
   * @param {number} index - The index value.
   * @return {object} - The return value.
   */
  getOutputByIndex(index) {
    return this.__outputs[index]
  }

  /**
   * The getOutput method.
   * @param {string} name - The name value.
   * @return {OperatorOutput} - The return value.
   */
  getOutput(name) {
    for (const o of this.__outputs) {
      if (o.getName() == name) return o
    }
  }

  /**
   * The __evalOutput method.
   * @param {any} cleanedParam - The cleanedParam value.
   * @private
   */
  __evalOutput(cleanedParam /* value, getter */) {
    for (const o of this.__outputs) {
      o.removeCleanerFn(this.__evalOutput)
    }
    this.evaluate()

    // Why does the cleaner need to return a value?
    // Usually operators are connected to multiple outputs.
    // return getter(1);
  }

  /**
   * The evaluate method.
   * Computes the values of each of the outputs based on the values of the inputs
   * and the values of outputs with mode OP_READ_WRITE.
   * This method must be implemented by all Operators.
   */
  evaluate() {
    throw new Error('Not yet implemented')
  }

  /**
   * The setValue method.
   * Note: FIXME Sometimes outputs are used in places like statemachines,
   * where we would want the change to cause an event.
   * Note: when a user sets a parameter value that is being driven by
   * an operator, the operator can propagate the value back up the chain
   * to its inputs.
   * @param {any} value - The value param.
   * @return {number} - Returns the number of outputs.
   */
  setValue(value, output) {
    // TODO: Implement me for custom manipulations.
    return value
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   * @return {object} - Returns the json object.
   */
  toJSON(context, flags) {
    const j = super.toJSON(context, flags)
    j.type = sgFactory.getClassName(this)

    const inputs = []
    for (const input of this.__inputs) {
      inputs.push(input.toJSON(context, flags))
    }
    j.inputs = inputs

    const outputs = []
    for (const output of this.__outputs) {
      outputs.push(output.toJSON(context, flags))
    }
    j.outputs = outputs
    return j
  }

  /**
   * The fromJSON method decodes a json object for this type.
   *
   * @param {object} j - The json object this item must decode.
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   */
  fromJSON(j, context, flags) {
    super.fromJSON(j, context, flags)

    if (j.inputs) {
      for (let i = 0; i < this.__inputs.length; i++) {
        const output = this.__inputs[i]
        output.fromJSON(j.inputs[i], context)
      }
    }
    if (j.outputs) {
      for (let i = 0; i < this.__outputs.length; i++) {
        const output = this.__outputs[i]
        output.fromJSON(j.outputs[i], context)
      }
    }
    // Force an evaluation of the operator as soon as loading is done.
    // context.addPLCB(() => {
    //   this.setDirty()
    // })
  }
}

export { Operator }
