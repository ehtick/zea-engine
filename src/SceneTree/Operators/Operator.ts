/* eslint-disable @typescript-eslint/no-explicit-any */

import { BaseItem } from '../BaseItem'
import { OperatorInput } from './OperatorInput'
import { OperatorOutput } from './OperatorOutput'

/**
 * Class representing an operator.
 *
 * @extends BaseItem
 */
class Operator extends BaseItem {
  __inputs: Map<string, OperatorInput>
  __outputs: Map<string, OperatorOutput>
  /**
   * Create an operator.
   * @param name - The name value.
   */
  constructor(name: string = '') {
    super(name)
    this.__inputs = new Map()
    this.__outputs = new Map()
  }

  /**
   * This method sets the state of the operator to dirty which propagates
   * to the outputs of this operator, and which may then propagate to other
   * operators. When the scene is cleaned, which usually is caused by rendering
   * then the chain of operators are cleaned by triggering evaluation.
   * @private
   */
  setDirty(): void {
    this.__outputs.forEach((output: OperatorOutput) => output.setDirty())
  }

  /**
   * This method can be overridden in derived classes
   * to perform general updates (see GLPass or BaseItem).
   *
   * @param event
   * @private
   */
  protected parameterValueChanged(event: Record<string, unknown>): void {
    super.parameterValueChanged(event)
    this.setDirty()
  }

  /**
   * The addInput method.
   * @param input - The name of the input, or the input object
   * @return - The return value.
   */
  addInput(input: OperatorInput): OperatorInput {
    if (typeof input == 'string') input = new OperatorInput(input)
    else if (!(input instanceof OperatorInput)) {
      throw new Error(`addInput only accepts string or OperatorInput`)
    }
    input.setOperator(this)
    this.__inputs.set(input.getName(), input)
    this.setDirty()
    return input
  }

  /**
   * The removeInput method.
   * @param input - The name of the input, or the input object
   */
  removeInput(input: OperatorInput): void {
    if (typeof input == 'string') input = this.getInput(input)
    if (!(input instanceof OperatorInput)) {
      throw new Error(`removeInput only accepts string or OperatorInput`)
    }
    if (input.getParam()) input.setParam(undefined)
    this.__inputs.delete(input.getName())
  }

  /**
   * Getter for the number of inputs in this operator.
   * @return - Returns the number of inputs.
   */
  getNumInputs(): number {
    return this.__inputs.size
  }

  /**
   * The getInputByIndex method.
   * @param index - The index value.
   * @return - The return value.
   */
  getInputByIndex(index: number): Record<string, any> {
    return Array.from(this.__inputs.values())[index]
  }

  /**
   * The getInput method.
   * @param name - The name value.
   * @return - The return value.
   */
  getInput(name: string): OperatorInput {
    const input = this.__inputs.get(name)
    if (!input) throw `Couldn't find an Input with the name of '${name}'`
    return input
  }

  /**
   * The addOutput method.
   * @param output - The name of the output, or the output object
   * @return - The return value.
   */
  addOutput(output: OperatorOutput | string): OperatorOutput {
    if (typeof output == 'string') output = new OperatorOutput(output)
    else if (!(output instanceof OperatorOutput)) {
      throw new Error(`addOutput only accepts string or OperatorOutput`)
    }
    output.setOperator(this)
    // if (this.getOutput(output.getName())) throw new Error(`Operator output already exists ${output.getName()}`)
    this.__outputs.set(output.getName(), output)
    this.setDirty()
    return output
  }

  /**
   * The removeOutput method.
   * @param output - The name of the output, or the output object
   */
  removeOutput(output: OperatorOutput | string): void {
    if (typeof output == 'string') output = this.getOutput(output)
    if (!(output instanceof OperatorOutput)) {
      throw new Error(`removeOutput only accepts string or OperatorInput`)
    }
    if (output.getParam()) output.setParam()
    this.__outputs.delete(output.getName())
  }

  /**
   * Getter for the number of outputs in this operator.
   * @return - Returns the number of outputs.
   */
  getNumOutputs(): number {
    return this.__outputs.size
  }

  /**
   * The getOutputByIndex method.
   * @param index - The index value.
   * @return - The return value.
   */
  getOutputByIndex(index: number): OperatorOutput {
    return Array.from(this.__outputs.values())[index]
  }

  /**
   * The getOutput method.
   * @param name - The name value.
   * @return - The return value.
   */
  getOutput(name: string): OperatorOutput {
    const output = this.__outputs.get(name)
    if (!output) throw new Error(`Couldn't find an Output with the name of '${name}'`)
    return output
  }

  /**
   * The evaluate method.
   * Computes the values of each of the outputs based on the values of the inputs
   * and the values of outputs with mode OP_READ_WRITE.
   * This method must be implemented by all Operators.
   */
  evaluate(): void {
    throw new Error('Not yet implemented')
  }

  /**
   * When the value on a Parameter is modified by a user by calling 'setValue,
   * then if any operators are bound, the value of the Parameter cannot be modified
   * directly as it is the result of a computation. Instead, the Parameter calls
   * 'backPropagateValue' on the Operator to cause the Operator to handle propagating
   * the value to one or more of its inputs.
   * to its inputs.
   * @param value - The value param.
   * @return - The modified value.
   */
  backPropagateValue(value: unknown): unknown {
    // TODO: Implement me for custom manipulations.
    return value
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param context - The context value.
   * @return - Returns the json object.
   */
  toJSON(context?: Record<string, any>): Record<string, any> {
    const j = super.toJSON(context)
    ;(j as any).type = this.getClassName()
    const inputs: any = []
    this.__inputs.forEach((input: any) => {
      inputs.push(input.toJSON(context))
    })
    ;(j as any).inputs = inputs
    const outputs: any = []
    this.__outputs.forEach((output: any) => {
      outputs.push(output.toJSON(context))
    })
    ;(j as any).outputs = outputs
    return j
  }

  /**
   * The fromJSON method decodes a json object for this type.
   *
   * @param j - The json object this item must decode.
   * @param context - The context value.
   */
  fromJSON(j: Record<string, any>, context?: Record<string, any>): void {
    super.fromJSON(j, context)
    if (j.inputs) {
      j.inputs.forEach((inputJson: any, index: any) => {
        let input
        if (inputJson.name) {
          input = this.getInput(inputJson.name)
          if (!input) {
            input = this.addInput(inputJson.name)
          }
        } else {
          input = this.getInputByIndex(index)
        }
        ;(input as OperatorInput).fromJSON(inputJson, context)
      })
    }
    if (j.outputs) {
      j.outputs.forEach((outputJson: any, index: any) => {
        let output
        if (outputJson.name) {
          output = this.getOutput(outputJson.name)
          if (!output) {
            output = this.addOutput(outputJson.name)
          }
        } else {
          output = this.getOutputByIndex(index)
        }
        ;(output as OperatorOutput).fromJSON(outputJson, context)
      })
    }
  }

  /**
   * The detach method.
   */
  detach(): void {
    this.__inputs.forEach((input: OperatorInput) => input.detach())
    this.__outputs.forEach((output: OperatorOutput) => output.detach())
  }

  /**
   * The reattach method.
   */
  reattach(): void {
    this.__inputs.forEach((input: OperatorInput) => input.reattach())
    this.__outputs.forEach((output: OperatorOutput) => output.reattach())
  }

  /**
   * The rebind method.
   */
  rebind(): void {
    this.__outputs.forEach((output: OperatorOutput) => output.rebind())
  }
}

export { Operator }
