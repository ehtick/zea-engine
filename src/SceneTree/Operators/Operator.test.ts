/* eslint-disable @typescript-eslint/no-explicit-any */
import { Operator } from './Operator'
import { NumberParameter } from '../Parameters/NumberParameter'
import { BaseItem } from '../BaseItem'
import { OperatorInput } from './OperatorInput'
import { OperatorOutput } from './OperatorOutput'
import { Registry } from '../../Registry'
import { OperatorOutputMode } from '../Parameters/OperatorOutputMode'

class AddFloatsOperator extends Operator {
  constructor(name?: string) {
    super(name)
    this.addInput(new OperatorInput('A'))
    this.addInput(new OperatorInput('B'))
    this.addOutput(new OperatorOutput('C'))
  }

  evaluate() {
    const a = this.getInput('A').getValue() as number
    const b = this.getInput('B').getValue() as number
    this.getOutput('C').setClean(a + b)
  }
}

Registry.register('AddFloatsOperator', AddFloatsOperator)

// Note: this operator modifies a value in the output attribute
// By reading and then changing. This feature allows us to combine operators
// to compute complex results. (See BoundingBox operators).
class ScaleFloatOperator extends Operator {
  constructor(name?: string) {
    super(name)
    this.addInput(new OperatorInput('ScaleValue'))
    this.addOutput(new OperatorOutput('Result', OperatorOutputMode.OP_READ_WRITE))
  }

  evaluate() {
    let scaleValue = 1.0
    const inParam = this.getInput('ScaleValue').getParam()
    if (inParam) {
      scaleValue = inParam.getValue() as number
    }
    // Read the value, modify and return.
    const value = this.getOutput('Result').getValue() as number
    this.getOutput('Result').setClean(value * scaleValue)
  }
}

Registry.register('ScaleFloatOperator', ScaleFloatOperator)

describe('Operator', () => {
  test('AddFloatsOperator', () => {
    const addOperator = new AddFloatsOperator()

    const aParam = new NumberParameter('A')
    const bParam = new NumberParameter('B')
    const myParam = new NumberParameter('MyParam')

    addOperator.getInput('A').setParam(aParam)
    addOperator.getInput('B').setParam(bParam)
    addOperator.getOutput('C').setParam(myParam)

    aParam.setValue(3)

    bParam.setValue(2.5)

    expect(myParam.isDirty()).toBe(true)

    expect(myParam.getValue() as number).toBe(5.5)

    expect(myParam.isDirty()).toBe(false)
  })

  test('ScaleFloatOperator', () => {
    const scaleOperator = new ScaleFloatOperator()

    const scaleParam = new NumberParameter('A', 2)
    const resultParam = new NumberParameter('MyParam', 3)

    scaleOperator.getInput('ScaleValue').setParam(scaleParam)
    scaleOperator.getOutput('Result').setParam(resultParam)

    expect(resultParam.isDirty()).toBe(true)

    expect(resultParam.getValue() as number).toBe(6)

    expect(resultParam.isDirty()).toBe(false)

    scaleParam.setValue(4)

    expect(resultParam.isDirty()).toBe(true)

    expect(resultParam.getValue() as number).toBe(24)

    expect(resultParam.isDirty()).toBe(false)
  })

  test('combining AddFloatsOperator and ScaleFloatOperator', () => {
    const addOperator = new AddFloatsOperator()
    const aParam = new NumberParameter('A', 2)
    const bParam = new NumberParameter('B', 3.5)
    const myParam = new NumberParameter('AddScaleFloatOperator-MyParam')

    addOperator.getInput('A').setParam(aParam)
    addOperator.getInput('B').setParam(bParam)
    addOperator.getOutput('C').setParam(myParam)

    const scaleOperator = new ScaleFloatOperator()
    const scaleParam = new NumberParameter('A', 2)

    scaleOperator.getInput('ScaleValue').setParam(scaleParam)
    scaleOperator.getOutput('Result').setParam(myParam)

    expect(myParam.isDirty()).toBe(true)

    expect(myParam.getValue()).toBe(11)

    expect(myParam.isDirty()).toBe(false)

    scaleParam.setValue(1)

    expect(myParam.isDirty()).toBe(true)

    expect(myParam.getValue()).toBe(5.5)

    expect(myParam.isDirty()).toBe(false)
  })

  test('dynamically changing inputs and outputs', () => {
    const operator = new Operator()
    const aParam = new NumberParameter('A', 2)
    const bParam = new NumberParameter('B', 3.5)
    const cParam = new NumberParameter('C')

    operator.addInput(new OperatorInput('A')).setParam(aParam)
    operator.addInput(new OperatorInput('B')).setParam(bParam)
    operator.addOutput(new OperatorOutput('C')).setParam(cParam)

    expect(operator.getNumInputs()).toBe(2)

    expect(operator.getNumOutputs()).toBe(1)

    expect(aParam.isDirty()).toBe(false)

    expect(bParam.isDirty()).toBe(false)

    expect(cParam.isDirty()).toBe(true)

    operator.getInput('A').setParam(null)
    operator.getInput('B').setParam(null)
    operator.getOutput('C').setParam(null)

    expect(aParam.isDirty()).toBe(false)

    expect(bParam.isDirty()).toBe(false)

    expect(cParam.isDirty()).toBe(false)

    operator.removeInput(operator.getInput('A'))
    operator.removeInput(operator.getInput('B'))
    operator.removeOutput(operator.getOutput('C'))

    expect(operator.getNumInputs()).toBe(0)

    expect(operator.getNumOutputs()).toBe(0)

    operator.addInput(new OperatorInput('A')).setParam(aParam)
    operator.addInput(new OperatorInput('B')).setParam(bParam)
    operator.addOutput(new OperatorOutput('C')).setParam(cParam)

    expect(aParam.isDirty()).toBe(false)

    expect(bParam.isDirty()).toBe(false)

    expect(cParam.isDirty()).toBe(true)

    operator.removeInput(operator.getInput('A'))
    operator.removeInput(operator.getInput('B'))
    operator.removeOutput(operator.getOutput('C'))

    expect(aParam.isDirty()).toBe(false)

    expect(bParam.isDirty()).toBe(false)

    expect(cParam.isDirty()).toBe(false)
  })

  class SetFloatOperator extends Operator {
    value: any
    constructor(name: any, value: any) {
      super(name)
      this.value = value
      this.addOutput(new OperatorOutput('Output', OperatorOutputMode.OP_WRITE))
    }

    evaluate() {
      this.getOutput('Output').setClean(this.value)
    }
  }

  class ScaleFloatsOperator extends Operator {
    constructor(name: string) {
      super(name)
      this.addInput(new OperatorInput('ScaleValue'))
      this.addOutput(new OperatorOutput('OutputA', OperatorOutputMode.OP_READ_WRITE))
      this.addOutput(new OperatorOutput('OutputB', OperatorOutputMode.OP_READ_WRITE))
    }

    evaluate() {
      let scaleValue = 2.0
      const inParam = this.getInput('ScaleValue').getParam()
      if (inParam) {
        scaleValue = inParam.getValue()
      }
      // Read the value, modify and return both values
      const valueA = this.getOutput('OutputA').getValue() as number
      this.getOutput('OutputA').setClean(valueA * scaleValue)
      const valueB = this.getOutput('OutputB').getValue() as number
      this.getOutput('OutputB').setClean(valueB * scaleValue)

      // this.emit('evaluated')
    }
  }

  test('horizontal dirty propagation', () => {
    const aParam = new NumberParameter('A')
    const bParam = new NumberParameter('B')
    const cParam = new NumberParameter('C')
    const scaleABParam = new NumberParameter('scaleABParam', 2)
    const scaleBCParam = new NumberParameter('scaleBCParam', 2)

    // In the following configuration, we have 3 parameters being driven by
    // 5 different operators. The 'setA/B/C' operators initialize the values to 2
    // Then the scale operators scale that value using the 'scaleAB' and 'scaleBC' values.
    //     Param A: > ['setA', 'scaleAB']
    //     Param B: > ['setB', 'scaleAB', 'scaleBC']
    //     Param C: > ['setC', 'scaleBC']
    // Parameter 'B' is modified by both scale operators.
    const setA = new SetFloatOperator('setA', 2)
    setA.getOutput('Output').setParam(aParam)
    const setB = new SetFloatOperator('setB', 2)
    setB.getOutput('Output').setParam(bParam)
    const setC = new SetFloatOperator('setC', 2)
    setC.getOutput('Output').setParam(cParam)

    const scaleAB = new ScaleFloatsOperator('scaleAB')
    scaleAB.getOutput('OutputA').setParam(aParam)
    scaleAB.getOutput('OutputB').setParam(bParam)
    scaleAB.getInput('ScaleValue').setParam(scaleABParam)

    const scaleBC = new ScaleFloatsOperator('scaleBC')
    scaleBC.getOutput('OutputA').setParam(bParam)
    scaleBC.getOutput('OutputB').setParam(cParam)
    scaleBC.getInput('ScaleValue').setParam(scaleBCParam)

    expect(aParam.isDirty()).toBe(true)

    expect(bParam.isDirty()).toBe(true)

    expect(cParam.isDirty()).toBe(true)

    expect(aParam.getValue()).toBe(4)

    expect(bParam.getValue()).toBe(8)

    expect(cParam.getValue()).toBe(4)

    expect(aParam.isDirty()).toBe(false)

    expect(bParam.isDirty()).toBe(false)

    expect(cParam.isDirty()).toBe(false)

    // Now we do something interesting.
    // We modify scaleAB, which will dirty A and B.
    // However, because ParamB becomes dirty, and it must dirty its
    // entire stack up to the last 'OP_WRITE' connected output,
    // then it also propagates dirty up to scaleBC operator, which then
    // propagates down to all its outputs, which includes C

    scaleABParam.setValue(3)

    expect(aParam.isDirty()).toBe(true)

    expect(bParam.isDirty()).toBe(true)

    expect(cParam.isDirty()).toBe(true) // cParam becomes dirty because bParam becomes dirty

    expect(aParam.getValue()).toBe(6) // 3 * 2

    expect(bParam.getValue()).toBe(12) // (3 * 2) * 2

    expect(cParam.getValue()).toBe(4) // 2 * 2

    expect(aParam.isDirty()).toBe(false)

    expect(bParam.isDirty()).toBe(false)

    expect(cParam.isDirty()).toBe(false)

    scaleBCParam.setValue(3)

    expect(aParam.isDirty()).toBe(true)

    expect(bParam.isDirty()).toBe(true)

    expect(cParam.isDirty()).toBe(true)

    expect(aParam.getValue()).toBe(6) // 3 * 2

    expect(bParam.getValue()).toBe(18) // (3 * 2) * 2

    expect(cParam.getValue()).toBe(6) // 3 * 2

    expect(aParam.isDirty()).toBe(false)

    expect(bParam.isDirty()).toBe(false)

    expect(cParam.isDirty()).toBe(false)
  })

  test('creating an cyclic dependency caused by mixing OP_READ_WRITE layering', () => {
    const aParam = new NumberParameter('A')
    const bParam = new NumberParameter('B')
    const scaleABParam1 = new NumberParameter('scaleABParam1', 2)
    const scaleABParam2 = new NumberParameter('scaleABParam2', 2)

    const setA = new SetFloatOperator('setA', 2)
    setA.getOutput('Output').setParam(aParam)
    const setB = new SetFloatOperator('setB', 2)
    setB.getOutput('Output').setParam(bParam)

    const scaleAB1 = new ScaleFloatsOperator('scaleAB1')
    scaleAB1.getInput('ScaleValue').setParam(scaleABParam1)
    const scaleAB2 = new ScaleFloatsOperator('scaleAB2')
    scaleAB2.getInput('ScaleValue').setParam(scaleABParam2)

    // Now we are going to mix up the bindings.
    // In theory, we should see an operator writing
    // to an output our of schedule. Meaning that
    // scaleAB1 should write the value of

    // Bind aParam: > ['scaleAB1', 'scaleAB2']
    scaleAB1.getOutput('OutputA').setParam(aParam)
    scaleAB2.getOutput('OutputA').setParam(aParam)

    // Bind bParam: > ['scaleAB2', 'scaleAB1']
    scaleAB2.getOutput('OutputB').setParam(bParam)
    scaleAB1.getOutput('OutputB').setParam(bParam)

    expect(aParam.isDirty()).toBe(true)

    expect(bParam.isDirty()).toBe(true)

    // This throws because we cannot evaluated scaleAB1 because its
    // input value depends on the value of scaleAB2, whose input also
    // depends on the output of scaleAB1

    expect(bParam.getValue).toThrow()
  })

  test('rebind to fix a cyclic dependency caused by mixing OP_READ_WRITE layering', () => {
    const aParam = new NumberParameter('A')
    const bParam = new NumberParameter('B')
    const scaleABParam1 = new NumberParameter('scaleABParam1', 2)
    const scaleABParam2 = new NumberParameter('scaleABParam2', 2)

    const setA = new SetFloatOperator('setA', 2)
    setA.getOutput('Output').setParam(aParam)
    const setB = new SetFloatOperator('setB', 2)
    setB.getOutput('Output').setParam(bParam)

    const scaleAB1 = new ScaleFloatsOperator('scaleAB1')
    scaleAB1.getInput('ScaleValue').setParam(scaleABParam1)
    const scaleAB2 = new ScaleFloatsOperator('scaleAB2')
    scaleAB2.getInput('ScaleValue').setParam(scaleABParam2)

    // Now we are going to mix up the bindings.
    // In theory, we should see an operator writing
    // to an output our of schedule. Meaning that
    // scaleAB1 should write the value of

    // Bind aParam: > ['scaleAB1', 'scaleAB2']
    scaleAB1.getOutput('OutputA').setParam(aParam)
    scaleAB2.getOutput('OutputA').setParam(aParam)

    // Bind bParam: > ['scaleAB2', 'scaleAB1']
    scaleAB2.getOutput('OutputB').setParam(bParam)
    scaleAB1.getOutput('OutputB').setParam(bParam)

    expect(aParam.isDirty()).toBe(true)

    expect(bParam.isDirty()).toBe(true)

    // This throws because we cannot evaluated scaleAB1 because its
    // input value depends on the value of scaleAB2, whose input also
    // depends on the output of scaleAB1
    // expect(bParam.getValue).toThrow()

    // Rebind forces the operators to remove and re-add bindings, which flattens the bindings and fixes the problem.
    scaleAB1.rebind()
    scaleAB2.rebind()

    scaleABParam1.setValue(3)

    expect(aParam.getValue()).toBe(12) // (3 * 2) * 2

    expect(bParam.getValue()).toBe(12) // (3 * 2) * 2
  })

  class DoubleFloatsOperator extends Operator {
    constructor(name?: any) {
      super(name)
      this.addInput(new OperatorInput('in'))
      this.addOutput(new OperatorOutput('out', OperatorOutputMode.OP_WRITE))
    }

    evaluate() {
      const a = this.getInput('in').getValue() as number
      this.getOutput('out').setClean(a * 2)
    }
  }
  class MultiOutputOperator extends Operator {
    constructor(name?: any) {
      super(name)
      this.addInput(new OperatorInput('in'))
      this.addOutput(new OperatorOutput('out0', OperatorOutputMode.OP_WRITE))
      this.addOutput(new OperatorOutput('out1', OperatorOutputMode.OP_WRITE))
      this.addOutput(new OperatorOutput('out2', OperatorOutputMode.OP_WRITE))
    }

    evaluate() {
      const value = this.getInput('in').getParam().getValue()
      this.getOutput('out0').setClean(value)
      this.getOutput('out1').setClean(value)
      this.getOutput('out2').setClean(value)
    }
  }

  test('layering multi-out on a chain', () => {
    const inAParam = new NumberParameter('inA')
    const aParam = new NumberParameter('A')
    const bParam = new NumberParameter('B')
    const cParam = new NumberParameter('B')

    const doubleOperator0 = new DoubleFloatsOperator()
    doubleOperator0.getInput('in').setParam(inAParam)
    doubleOperator0.getOutput('out').setParam(aParam)

    const doubleOperator1 = new DoubleFloatsOperator()
    doubleOperator1.getInput('in').setParam(aParam)
    doubleOperator1.getOutput('out').setParam(bParam)

    const doubleOperator2 = new DoubleFloatsOperator()
    doubleOperator2.getInput('in').setParam(bParam)
    doubleOperator2.getOutput('out').setParam(cParam)

    inAParam.setValue(2)

    aParam.setValue(2.5)

    expect(aParam.getValue()).toBe(4)

    expect(bParam.getValue()).toBe(8)

    expect(cParam.getValue()).toBe(16)

    expect(cParam.isDirty()).toBe(false)

    const multiOutputOperator = new MultiOutputOperator()
    const inBParam = new NumberParameter('inB', 3)
    multiOutputOperator.getInput('in').setParam(inBParam)
    multiOutputOperator.getOutput('out0').setParam(aParam)
    multiOutputOperator.getOutput('out1').setParam(bParam)
    multiOutputOperator.getOutput('out2').setParam(cParam)

    expect(aParam.getValue()).toBe(3)

    expect(bParam.getValue()).toBe(3)

    expect(cParam.getValue()).toBe(3)

    expect(cParam.isDirty()).toBe(false)

    // Now dirty the input to the chain.
    // Because the multi-out layers on top using OP_WRITE outputs,
    // it should prevent the chain from becoming dirty.

    inAParam.setValue(4)

    expect(cParam.isDirty()).toBe(false)

    expect(aParam.getValue()).toBe(3)

    expect(bParam.getValue()).toBe(3)

    expect(cParam.getValue()).toBe(3)
  })

  test('save to JSON (serialization).', () => {
    const addOperator = new AddFloatsOperator()
    const parameterOwner = new BaseItem('Foo')
    const aParam = parameterOwner.addParameter(new NumberParameter('A'))
    const bParam = parameterOwner.addParameter(new NumberParameter('B'))
    const myParam = parameterOwner.addParameter(new NumberParameter('MyParam'))

    addOperator.getInput('A').setParam(aParam)
    addOperator.getInput('B').setParam(bParam)
    addOperator.getOutput('C').setParam(myParam)

    aParam.setValue(3)
    bParam.setValue(2.5)

    expect(JSON.stringify(addOperator.toJSON())).toMatchSnapshot()
  })

  test('load from JSON (serialization).', () => {
    const parameterOwner = new BaseItem('Foo')
    parameterOwner.addParameter(new NumberParameter('A'))
    parameterOwner.addParameter(new NumberParameter('B'))
    parameterOwner.addParameter(new NumberParameter('MyParam'))

    const addOperator = new AddFloatsOperator()
    const input = {
      name: '',
      type: 'AddFloatsOperator',
      inputs: [
        { name: 'A', paramPath: ['Foo', 'A'] },
        { name: 'B', paramPath: ['Foo', 'B'] },
      ],
      outputs: [{ name: 'C', paramPath: ['Foo', 'MyParam'], paramBindIndex: 0 }],
    }
    addOperator.fromJSON(input, {
      resolvePath: (path: any, cb: any) => {
        cb(parameterOwner.resolvePath(path))
      },
    })

    parameterOwner.getParameter('A').setValue(3)
    parameterOwner.getParameter('B').setValue(2.5)

    expect(parameterOwner.getParameter('MyParam').isDirty()).toBe(true)

    expect(parameterOwner.getParameter('MyParam').getValue()).toBe(5.5)

    expect(parameterOwner.getParameter('MyParam').isDirty()).toBe(false)
  })
})
