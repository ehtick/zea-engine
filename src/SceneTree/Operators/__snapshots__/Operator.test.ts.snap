// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operator save to JSON (serialization). 1`] = `"{\\"type\\":\\"AddFloatsOperator\\",\\"name\\":\\"\\",\\"inputs\\":[{\\"name\\":\\"A\\",\\"paramPath\\":[\\"Foo\\",\\"A\\"]},{\\"name\\":\\"B\\",\\"paramPath\\":[\\"Foo\\",\\"B\\"]}],\\"outputs\\":[{\\"name\\":\\"C\\",\\"paramPath\\":[\\"Foo\\",\\"MyParam\\"],\\"paramBindIndex\\":0}]}"`;
